circuits/circuits/claiming/challenge.circom

```diff
@@ -0,0 +1,144 @@
+pragma circom 2.1.5;
+
+include "../node_modules/circomlib/circuits/poseidon.circom"; // TODO: consider Poseidon2
+include "../node_modules/circomlib/circuits/comparators.circom";
+
+// Every leaf in the filtered tree is either equal to the same node in the claim tree, or it is zero
+// The leaf is 0 if it is invalid or it is beaten by a challenger
+template FilteredLeafIsValid(claim_levels, state_levels, upper_state_levels, i) {
+    // Merkle roots
+    signal input states_history_root;
+    signal input claim_root;
+    signal input filtered_root;
+
+    // Defender
+    signal input defender[3];
+    signal input defender_state_PathElements[state_levels];
+    signal input defender_state_PathIndices[state_levels];
+    signal input defender_claim_PathElements[claim_levels];
+
+    // Challenger
+    signal input challenger_index;
+    signal input challenger[3];
+    signal input challenger_state_PathElements[state_levels];
+    signal input challenger_state_PathIndices[state_levels];
+    signal input challenger_claim_PathElements[claim_levels];
+
+    // Results
+    signal input filtered_pathElements[claim_levels];
+
+    signal claim_is_valid <== ClaimIsValid(claim_levels, state_levels)(
+        i,
+        defender,
+        defender_state_PathElements,
+        defender_state_PathIndices,
+        defender_claim_PathElements,
+        states_history_root,
+        claim_root
+    );
+
+    // Make sure the challenger is a real leaf in the claim tree with a valid proof of ownership
+    // The prover can always find *some* challenger in the tree as long as there is at least one valid claim
+    signal challenge_is_valid <== ClaimIsValid(claim_levels, state_levels)(
+        challenger_index,
+        challenger,
+        challenger_state_PathElements,
+        challenger_state_PathIndices,
+        challenger_claim_PathElements,
+        states_history_root,
+        claim_root
+    );
+    challenge_is_valid === 1;
+
+    signal challenge_succeeds <== ChallengeSucceeds(state_levels, upper_state_levels)(
+        defender <== defender,
+        challenger <== challenger,
+        defender_state_indices <== defender_state_PathIndices,
+        challenger_state_indices <== challenger_state_PathIndices
+    );
+
+    // if the claim is valid and the challenge doesn't succeed, the leaves in the claim and filtered trees should be equal
+    signal should_be_equal <== claim_is_valid * (1 - challenge_succeeds);
+    signal claim_hash <== Poseidon(3)(defender);
+    signal filtered_leaf <== claim_hash * should_be_equal;
+    CheckMerkleProofStrict(claim_levels)(
+        filtered_leaf,
+        filtered_pathElements,
+        Num2Bits(claim_levels)(in <== i),
+        filtered_root
+    );
+}
+
+// Asserts that the claim is in the claim tree
+// Returns whether the claim is properly formed and is in the state history
+template ClaimIsValid(claim_levels, state_levels) {
```

Comment:
Ah OK fair point. I don't like `IsClaimValid` grammatically because it's phrased like a question, but `IsValidClaim` works, right? So `Is` basically means boolean/`0/1`?

Reply:
commented about this elsewhere too. but yeah i see `Is` as a boolean marker. no problem if this is different from your style.